Home Assistant templates

balloob opened the pull request:
Ever since jsonpath was implemented for MQTT sensors and switches, I wanted to make this available for all things. Then yesterday, someone on the mailing list asked for variables in notifications and then it struck me: templates will be the solution for everything.

Templates will:
- replace options like precision or correction in platforms like `arest`, `rest`, `mqtt`
- replace jsonpath for MQTT switches and sensors

This PR will add MVP Template support. On top of Jinja2 I've added the special `states` variable:
- Iterating `states` will yield each state sorted alphabetically by entity id
- Iterating `states.domain` will yield each state of that domain sorted alphabetically by entity id
- `states.sensor.temperature` returns state object for `sensor.temperature`

I've also added two methods `multiply` and `round`. Both which convert strings to numbers before doing their job.

Example templates and what they could result in:

| Template | Output |
| --- | --- |
| `{{ states.device_tracker.paulus.state }}` | home |
| `{% for state in states.sensor %}{{ state.entity_id }}={{ state.state }}, {% endfor %}` | senor.thermostat=24, sensor.humidity=40, |
| `{{ value['hello'][0] | multiply(10) | round(1) }}` | 10.2 |

[Jinja2 Template documentation](http://jinja.pocoo.org/docs/dev/templates/)

MVP will consist of:
- [x] Add template util methods + tests
- [x] Notification component should use templates + tests
- [x] Add template API endpoint `/api/template`
- [x] Replace jsonpath support in MQTT with templates

Version 2 (not this PR):
- Allow template inclusion from the config dir
- Add template support to rest, aRest
- Add template support to numeric state automation
- Add template automation rule.

balloob committed:
Add template support

balloob committed:
Add jinja2 to dependencies.

balloob committed:
Add multiply template filter

happyleavesaoc commented:
Can you expand on "Allow template inclusion from the config dir"?

balloob commented:
Jinja2 allows including other templates from your template ([docs](http://jinja.pocoo.org/docs/dev/templates/#include)):
```
{% include 'header.html' %}
```

These files would then point at a directory in your config dir.

balloob committed:
Add template parsing to notify

balloob committed:
Add template API endpoint

philipbl commented on homeassistant/util/template.py:
```diff
@@ -0,0 +1,67 @@
+def forgiving_round(value, precision=0):
+    """ Rounding method that accepts strings. """
+    try:
+        return int(value) if precision == 0 else round(float(value), precision)
```
Should this line be:

```python
return int(round(float(value))) if precision == 0 else round(float(value), precision)
```

balloob replied:
Fixing it right now.

balloob committed:
Fix template rounding

balloob committed:
Add variable support to template rendering

balloob committed:
Add value renderer helper method

balloob committed:
Migrate MQTT from jsonpath to templates

balloob commented:
As part of this MVP I've replaced the jsonpath support in the MQTT components sensor, switch and rollershutter with template support. Jsonpath was great but templates will give us even more options.

I want to deprecate all configuration options to mutate the value we get back from a service (for example precision and correction_factor) and move to a world with only `value_template`. I do not plan on removing them right away but want to add warnings first.

CC @mcdeck @sfam as I have changed code they contributed.

philipbl commented on homeassistant/util/template.py:
```diff
@@ -0,0 +1,86 @@
+    if variables is not None:
+        kwargs.update(variables)
+
+    return ENV.from_string(template, {
```
Should this be wrapped in some kind of exception handling? If the template string is incorrect, a `jinja2.exceptions.UndefinedError` is raised. This means anyone that uses the `render` or `render_with_possible_json_value` would have to import `jinja2.exceptions` if they want to handle the exception properly.

balloob replied:
I thought about it adding it to render. I didn't do it because variables
that do not exist will be replaced with empty strings. I don't know what I
would return as a fallback for this method, the unprocessed template?

For the render_with_possible_json_value method I think logging the error
and returning the passed in value would be a good use case.

On Fri, Dec 11, 2015 at 10:20 AM, Philip Lundrigan <notifications@github.com
> wrote:

> In homeassistant/util/template.py
> <https://github.com/balloob/home-assistant/pull/720#discussion_r47386199>:
>
> > +        'value': value
> > +    }
> > +    try:
> > +        variables['value_json'] = json.loads(value)
> > +    except ValueError:
> > +        pass
> > +
> > +    return render(hass, template, variables)
> > +
> > +
> > +def render(hass, template, variables=None, **kwargs):
> > +    """ Render given template. """
> > +    if variables is not None:
> > +        kwargs.update(variables)
> > +
> > +    return ENV.from_string(template, {
>
> Should this be wrapped in some kind of exception handling? If the template
> string is incorrect, a jinja2.exceptions.UndefinedError is raised. This
> means anyone that uses the render or render_with_possible_json_value
> would have to import jinja2.exceptions if they want to handle the
> exception properly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/720/files#r47386199>.
>



-- 
PaulusSchoutsen.nl
It's nice to be important but it's more important to be nice.

philipbl replied:
I'm not sure if I understand what you mean by "variables that do not exist will be replaced with empty strings". Where is that happening?

I am trying to implement the Rest sensor using the new templates. If a user provides `{{ value_json.something_that_doesnt_exist }}` or `{{  something_that_doesnt_exist }}` as the template, the `from_string` method will raise an exception. I think it is acceptable to raise an exception in this cause because there is nothing else to do. We just want the exception to be of our own type, and not `jinja2.exceptions.UndefinedError` so that not every module has to import jinja2.

I'm probably missing something.

balloob replied:
Hmm, the second template should not raise an exception. Check the test with
invalid json for example.

Yes, you're right we should wrap the exceptions.

On Fri, Dec 11, 2015, 12:33 Philip Lundrigan <notifications@github.com>
wrote:

> In homeassistant/util/template.py
> <https://github.com/balloob/home-assistant/pull/720#discussion_r47401827>:
>
> > +        'value': value
> > +    }
> > +    try:
> > +        variables['value_json'] = json.loads(value)
> > +    except ValueError:
> > +        pass
> > +
> > +    return render(hass, template, variables)
> > +
> > +
> > +def render(hass, template, variables=None, **kwargs):
> > +    """ Render given template. """
> > +    if variables is not None:
> > +        kwargs.update(variables)
> > +
> > +    return ENV.from_string(template, {
>
> I'm not sure if I understand what you mean by "variables that do not exist
> will be replaced with empty strings". Where is that happening?
>
> I am trying to implement the Rest sensor using the new templates. If a
> user provides {{ value_json.something_that_doesnt_exist }} or {{
> something_that_doesnt_exist }} as the template, the from_string method
> will raise an exception. I think it is acceptable to raise an exception in
> this cause because there is nothing else to do. We just want the exception
> to be of our own type, and not jinja2.exceptions.UndefinedError so that
> not every module has to import jinja2.
>
> I'm probably missing something.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balloob/home-assistant/pull/720/files#r47401827>.
>

philipbl replied:
You're right, the second template doesn't raise an exception. But this slight modification does: `{{ value_json.something_that_doesnt_exist | multiply(10) }}`.
